R/model_ivpin.R

```diff
@@ -0,0 +1,905 @@
+  Vb <- bucketdata$agg.bvol
+  Vs <- bucketdata$agg.svol
+
+  ivpin <- numeric(nrow(bucketdata))  # Ensure ivpin is initialized
+
+  for (i in start_index:nrow(bucketdata)) {
+    j <- i - samplength
+    parms <- rep(-Inf, 4)  # alpha, delta, mu, eps
+    
+    log_lik <- Inf
+    flag <- -Inf
+    
+    best_log_likelihood <- Inf
+    exit_flag <- FALSE
+    
+    if (i == start_index || i %% 10 == 0) {
+      for (alpha_init in seq(0.1, 0.9, by = 0.2)) {
+        for (delta_init in seq(0.1, 0.9, by = 0.2)) {
+          mu_init <- mean(informed_arrival[j:i]) / alpha_init
+          eps_init <- mean(abs(total_arrival_rate[j:i] - informed_arrival[j:i])) / 2
+          
+          initial_guess <- c(logit(max(min(alpha_init, 0.999), 0.001)), 
+                             logit(max(min(delta_init, 0.999), 0.001)), 
+                             sqrt(mu_init), 
+                             sqrt(eps_init))
+          tryCatch({
+            result <- optim(initial_guess, compute_log_likelihood, t = t[j:i], Vb = Vb[j:i], Vs = Vs[j:i], method = "BFGS")
+          }, error = function(e) {
+            conditionMessage(e)
+          })
+          
+          if (result$value < best_log_likelihood && is.finite(result$value)) {
+            best_params <- result$par
+            best_log_likelihood <- result$value
+            exit_flag <- TRUE
+          }
+        }
+      }
+    } else {
+      initial_guess <- c(logit(best_params[1]), logit(best_params[2]), sqrt(best_params[3]), sqrt(best_params[4]))
+      tryCatch({
+        result <- optim(initial_guess, compute_log_likelihood, t = t[j:i], Vb = Vb[j:i], Vs = Vs[j:i], method = "BFGS")
+      }, error = function(e) {
+        conditionMessage(e)
+      })
+      
+      if (result$value < best_log_likelihood && is.finite(result$value)) {
+        best_params <- result$par
+        exit_flag <- TRUE
+      } else {
+        for (alpha_init in seq(0.1, 0.9, by = 0.2)) {
+          for (delta_init in seq(0.1, 0.9, by = 0.2)) {
+            mu_init <- informed_arrival[i] / alpha_init
+            eps_init <- abs(total_arrival_rate[i] - informed_arrival[i]) / 2
+            initial_guess <- c(logit(max(min(alpha_init, 0.999), 0.001)), 
+                               logit(max(min(delta_init, 0.999), 0.001)), 
+                               sqrt(mu_init), 
+                               sqrt(eps_init))
+            tryCatch({
+              result <- optim(initial_guess, compute_log_likelihood, t = t[j:i], Vb = Vb[j:i], Vs = Vs[j:i], method = "BFGS")
+            }, error = function(e) {
+              conditionMessage(e)
+            })
+            if (result$value < best_log_likelihood && is.finite(result$value)) {
+              best_params <- result$par
+              best_log_likelihood <- result$value
+              exit_flag <- TRUE
+            }
+          }
+        }
+      }
+    }
+    
+    if (exists("best_params")) {
+      best_params[1:2] <- sigmoid(best_params[1:2])
```

Comment:
This is good! I understand much better the logic behind your code. However, I would like to know the logic behind the back and forth movement between `logit()` and `sigmoid()`; and `sqrt()` and `^2`. n